cou/apps/core.py

```diff
@@ -116,32 +144,42 @@ def _get_empty_hypervisor_step(self, unit: Unit) -> UnitUpgradeStep:
             coro=verify_empty_hypervisor(unit, self.model),
         )
 
-    def _get_enable_scheduler_step(self, unit: Unit) -> UnitUpgradeStep:
+    def _get_enable_scheduler_step(self, units: Optional[list[Unit]]) -> list[PostUpgradeStep]:
         """Get the step to enable the scheduler, so the unit can create new VMs.
 
-        :param unit: Unit to be enabled.
-        :type unit: Unit
-        :return: Step to enable the scheduler.
-        :rtype: UnitUpgradeStep
+        :param units: Units to be enabled.
+        :type units: Optional[list[Unit]]
+        :return: Steps to enable the scheduler on units
+        :rtype: list[PostUpgradeStep]
         """
-        return UnitUpgradeStep(
-            description=f"Enable nova-compute scheduler from unit: '{unit.name}'",
-            coro=self.model.run_action(
-                unit_name=unit.name, action_name="enable", raise_on_failure=True
-            ),
-        )
-
-    def _get_disable_scheduler_step(self, unit: Unit) -> UnitUpgradeStep:
-        """Get the step to disable the scheduler, so the unit cannot create new VMs.
-
-        :param unit: Unit to be disabled.
-        :type unit: Unit
-        :return: Step to disable the scheduler.
-        :rtype: UnitUpgradeStep
+        if not units:
+            units = list(self.units.values())
+        return [
+            PostUpgradeStep(
+                description=f"Enable nova-compute scheduler from unit: '{unit.name}'",
+                coro=self.model.run_action(
+                    unit_name=unit.name, action_name="enable", raise_on_failure=True
+                ),
+            )
+            for unit in units
+        ]
+
+    def _get_disable_scheduler_step(self, units: Optional[list[Unit]]) -> list[PreUpgradeStep]:
+        """Get the step to disable the scheduler,  so the unit cannot create new VMs.
+
+        :param units: Units to be disabled.
+        :type units:  Optional[list[Unit]]
+        :return: Steps to disable the scheduler on units
+        :rtype: list[PreUpgradeStep]
         """
-        return UnitUpgradeStep(
-            description=f"Disable nova-compute scheduler from unit: '{unit.name}'",
-            coro=self.model.run_action(
-                unit_name=unit.name, action_name="disable", raise_on_failure=True
-            ),
-        )
+        if not units:
+            units = list(self.units.values())
+        return [
+            PreUpgradeStep(
+                description=f"Disable nova-compute scheduler from unit: '{unit.name}'",
+                coro=self.model.run_action(
+                    unit_name=unit.name, action_name="disable", raise_on_failure=True
```

Comment:
not-blocking: What if we use one functions with `enable=True` for enabling (run enable action) and `enable=False` for disabling (run disable action).

Reply:
Good point, resolving. :+1: